### PR TITLE
webpack

### PR DESCRIPTION
**Pull request by jdfreder**
Allows the notebook to use NPM packages, commonjs, and transpilers if desired, while maintaining support for require.js and adding shims for old-style requirejs path based imports.  Suggests an alternative, explicit export approach for requirejs module loading.  

/me "That's one small leap for man, one giant leap for Jovyans"

### CODE REVIEW

**Comment by jdfreder** [comment]
This is ready!

@Carreau this retains all of the require.js behavior, but as we start adding dependencies on true npm packages, it will make that more difficult.  I'm thinking about adding an index.js or change namespace.js so that it does and exposes imports to various useful modules.  Then, instead of telling people to 

```js
define(['path/to/module'], function(module) {
})
``` 

we would tell them to 

```js
define(function(require) {
var module = require('notebook')['path/to/module'];
});
```

or  

```js
define(function(require) {
var module = require('notebook')['modulename'];
});
```

(in both cases notebook would be a require js config alias that pointed to a webpacked file somewhere)  

This way people have a way of importing the webpacked version of the module.

**Comment by jdfreder** [comment]
@Carreau maybe we can schedule a public, recorded meeting to discuss, it may be hard to explain everything without being extremely wordy. :stuck_out_tongue:

**Comment by ellisonbg** [comment]
Yeah, let's definitely have a meeting about this. With all the new code we
will have in JupyteLab, I think we need to be extra careful to keep the
other pages working the same for 5.0.

On Fri, Feb 5, 2016 at 5:43 PM, Jonathan Frederic <notifications@github.com>
wrote:

> @Carreau <https://github.com/Carreau> maybe we can schedule a public,
> recorded meeting to discuss, it may be hard to explain everything without
> being extremely wordy. [image: :stuck_out_tongue:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/1047#issuecomment-180652708>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by jdfreder** [comment]
This should not affect existing code because it leaves requirejs intact.  Existing code only becomes unloadable via restful path when/if an npm module is used.  It is in that case that existing code must be loaded via a webpacked bundle.

**Comment by Carreau** [comment]
1) I'm not unhappy because the watch things rebuild everything fast ! Thanks. 
2) It breaks IBM's guys `urth` as it cannot require jQuery:

```
utils.js:49 Failed to load extension: ["jquery"] Error: Script error for "jquery", needed by: nbextensions/urth_dash_js/notebook/deploy-menu/deploy-menu, nbextensions/widgets/notebook/js/widgetarea, nbextensions/urth_dash_js/notebook/dashboard-view/polymer-support(…)(anonymous function) @ utils.js:49
```

I'll comment inline, but thanks !

File: notebook/notebookapp.py [hunk 1]
```diff
@@ -399,11 +398,6 @@ def _log_format_default(self):
         """override default log format to include time"""
         return u"%(color)s[%(levelname)1.1s %(asctime)s.%(msecs).03d %(name)s]%(end_color)s %(message)s"
 
-    ignore_minified_js = Bool(False,
-            config=True,
-            help='Use minified JS file or not, mainly use during dev to avoid JS recompilation', 
-            )
-
```
Comment (Carreau) [inline notebook/notebookapp.py]: Keep this, emit a DeprecationWarning + a log warning message that it is deprecated, does not work and will be remove in 6.0, we have some instruction in the wild that say how to use it. 

I'm fine making it even refuse to start for after 5.0 but the during 5.0.dev transition, we can be explicit.

File: notebook/templates/page.html [hunk 2]
```diff
@@ -24,25 +29,25 @@
...
-            typeahead: 'components/jquery-typeahead/dist/jquery.typeahead'
+
+            // Anything jQuery related is loaded globally because it will
+            // pollute the global object regardless and we want to avoid one
+            // jQuery thinger from clompering another.
+            jquery: '//deprecated//',
+            bootstrap: '//deprecated//',
+            bootstraptour: '//deprecated//',
+            'jquery-ui': '//deprecated//',
+            typeahead: '//deprecated//'
```
Comment (Carreau) [inline notebook/templates/page.html]: Don't completely do that now, that will piss a lot of people bing on master.
Can you find a way if these are required to trigger a warning ?
Accessing from the global namespace in the few things you require seem to also be bad.

Comment (jdfreder) [inline notebook/templates/page.html]: Yeah, I agree requiring from the global namespace is bad, unfortunately jquery things insist on being global.

But yeah, I can totally make the requirements continue to work, and just console.warn a deprecation warning

File: notebook/notebookapp.py (hunk 1 above)
Comment (jdfreder) [inline notebook/notebookapp.py]: Good point!

**Comment by jdfreder** [comment]
> It breaks IBM's guys urth as it cannot require jQuery:

Ah yeah, I'll make jquery requirejs import warn.

**Comment by parente** [comment]
@Carreau thanks for testing this. It looks like you've got a slightly older build of the dashboards stuff installed, but I'm sure the latest will break too.

If I'm reading the diff right, the libraries will still be available in notebook "classic", but they just won't be accessible in the same way in 5.x and eventually will go away completely in 6.x.

It's understandable that for 5.0, extensions that are relying on these libs will need to change how they access them. For the 4.x series of code, though, it would be nice if, at worst, a deprecation message is logged but the libraries are still available in the same manner. Otherwise, any extensions / custom.js modifications that were depending on them will need to deal with shipping the lib in >=4.2 and using the notebook provided version in <=4.1.

**Comment by jdfreder** [comment]
> It's understandable that for 5.0, extensions that are relying on these libs will need to change how they access them. For the 4.x series of code, though, it would be nice if, at worst, a deprecation message is logged but the libraries are still available in the same manner. Otherwise, any extensions / custom.js modifications that were depending on them will need to deal with shipping the lib in >=4.2 and using the notebook provided version in <=4.1.

I totally agree, I will make sure they can be accessed the same way with deprecation warnings.  To me, that means we still do this, but we'll want to be careful to not use names that can't be imported by requirejs.

File: notebook/templates/page.html (hunk 2 above)
Comment (jdfreder) [inline notebook/templates/page.html]: :heavy_check_mark:

**Comment by jdfreder** [comment]
> For the 4.x series of code, though, it would be nice if, at worst, a deprecation message is logged but the libraries are still available in the same manner. Otherwise, any extensions / custom.js modifications that were depending on them will need to deal with shipping the lib in >=4.2 and using the notebook provided version in <=4.1.

:heavy_check_mark:

**Comment by jdfreder** [comment]
> Otherwise, any extensions / custom.js modifications that were depending on them will need to deal with shipping the lib in >=4.2 and using the notebook provided version in <=4.1.

Oh, this is target for 5.0.  You won't have to worry about this and 4.x.

File: notebook/notebookapp.py (hunk 1 above)
Comment (jdfreder) [inline notebook/notebookapp.py]: :heavy_check_mark:

**Comment by jdfreder** [comment]
Comments address and rebased!

**Comment by jdfreder** [comment]
I found a path forward that will make everything just work!  Even using requirejs to load npm dependent modules the old way !!  Of course this would be a deprecated behavior...  Working on these alterations now...

**Comment by jdfreder** [comment]
Total success, dev console proof - input: 

```
require(['notebook/js/cell'], function(x) { console.log(x); });
```

output:  

```
deprecated-imports.js:73 Direct module require deprecated.  Instead of using `define(['notebook/js/cell'], function(module) {});`, use `define(['notebook'], function(notebookApp) { var module =  notebookApp['notebook/js/cell']});`
VM969:2 Object {}
```

**Comment by rgbkrk** [comment]
Very awesome, excellent work Jon. Now refactoring bits of the current notebook and bringing in libraries to replace tiny components will be much easier.

**Comment by jdfreder** [comment]
Kind of crazy how it's done: requirejs bundles -> "deprecated-imports.js" -> "built/index.js".

We can later control what APIs we want exposed to the user explicitly in index.js (which builds to built/index.js)

**Comment by jdfreder** [comment]
> Very awesome, excellent work Jon. Now refactoring bits of the current notebook and bringing in libraries to replace tiny components will be much easier.

Thanks!

**Comment by parente** [comment]
> Total success, dev console proof - input:

That there is a feat of backward compatibility engineering. Impressive!

**Comment by jdfreder** [comment]
Need to trigger a Travis rebuild -

**Comment by jdfreder** [comment]
@Carreau check out ae26219 , it adds an info log statement like we talked about after the dev meeting. :smile:

**Comment by willingc** [comment]
@jdfreder Minor wording suggestions to log info message in https://github.com/jdfreder/notebook/commit/ae262191bb6edc89b4da9de34c9ec0e6ae479603

**Comment by jdfreder** [comment]
> @jdfreder Minor wording suggestions to log info message in jdfreder@ae26219

:heavy_check_mark: thanks!

**Comment by jdfreder** [comment]
Marking this as WIP because a bug cropped up and I'm trying to squish it before merge...  Also Travis.

**Comment by jdfreder** [comment]
Woot, okay this is good to go now!! 

I even added unit tests to make sure the old style require js imports continue to work.  My thinking is that these tests will make sure we don't break compatibility as we pull in npm packages.

**Comment by minrk** [comment]
That's awesome, @jdfreder!

**Comment by willingc** [comment]
@jdfreder :musical_keyboard: :notes: :guitar: Nice!

**Comment by minrk** [comment]
Login and logout pages don't seem to work. I'm guessing their main scripts didn't get the necessary updates:

screen shot 2016-02-11 at 15 29 12

**Comment by jdfreder** [comment]
> Login and logout pages don't seem to work. I'm guessing their main scripts didn't get the necessary updates:

I had the auth/main disable, because I assumed the main files weren't being loaded directly.  If you look at 801b158e9e , you'll see I uncommented the lines.  Should work now :)

**Comment by jdfreder** [comment]
I can confirm it works now, don't forget to run `npm run build` after checking out the changes

**Comment by minrk** [comment]
This is an interesting commit:

> 	Merge branch 'webpack' of github.com:jdfreder/notebook into webpack (d0da04f)

Do you know why that happened?

**Comment by minrk** [comment]
And login/logout are fixed, thanks!

**Comment by jdfreder** [comment]
> This is an interesting commit:
> 
> Merge branch 'webpack' of github.com:jdfreder/notebook into webpack (d0da04f)
> Do you know why that happened?

Yeah, I made the mistake of making changes on another computer which was behind my remote.  It asked me to pull, I said sure why not.  I'll squash it!  

...

Okay, squashed

**Comment by minrk** [comment]
...now this doesn't merge cleanly. Sorry!

**Comment by jdfreder** [comment]
Rebased- I'd like to avoid rebasing too many times if possible, is there consensus that we should move forward with the approach in this PR?  Should I ping additional people or the mailing list?

**Comment by Carreau** [comment]
> I'd like to avoid rebasing too many times if possible,

I understand. Sorry I wont have much time to review/try before next week. 

> Should I ping additional people or the mailing list?

Yes, I think that would be helpful, especially since it will likely break the notebook from people pulling from master.

**Comment by jdfreder** [comment]
Okay, I just sent a large email to the mailing list and pointed people to this PR.

**Comment by rgbkrk** [comment]
Great email wrapup on the mailing list! :rocket:

**Comment by jdfreder** [comment]
I gave @ellisonbg and @willingc an in-person overview

**Comment by willingc** [comment]
```
Installing collected packages: notebook
  Running setup.py develop for notebook
    Complete output from command /Users/carol/.virtualenvs/mypy3/bin/python3.5 -c "import setuptools, tokenize;__file__='/Users/carol/Projects/jupyter-world/notebook/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" develop --no-deps:
    running develop
    running jsversion
    running css
    running jsdeps
    bower dependencies up to date
    > lessc --source-map --include-path=/Users/carol/Projects/jupyter-world/notebook/notebook/static /Users/carol/Projects/jupyter-world/notebook/notebook/static/style/ipython.less /Users/carol/Projects/jupyter-world/notebook/notebook/static/style/ipython.min.css
    > lessc --source-map --include-path=/Users/carol/Projects/jupyter-world/notebook/notebook/static /Users/carol/Projects/jupyter-world/notebook/notebook/static/style/style.less /Users/carol/Projects/jupyter-world/notebook/notebook/static/style/style.min.css
    running js
    /Users/carol/.virtualenvs/mypy3/lib/python3.5/site-packages/setuptools/dist.py:285: UserWarning: Normalizing '5.0.0.dev' to '5.0.0.dev0'
      normalized_version,
    Traceback (most recent call last):
      File "/Users/carol/Projects/jupyter-world/notebook/setupbase.py", line 538, in run
        self.distribution.run_command('js')
      File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/Users/carol/Projects/jupyter-world/notebook/setupbase.py", line 484, in run
        pool.map(self.build_main, self.apps)
      File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/multiprocessing/pool.py", line 260, in map
        return self._map_async(func, iterable, mapstar, chunksize).get()
      File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/multiprocessing/pool.py", line 608, in get
        raise self._value
      File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/multiprocessing/pool.py", line 119, in worker
        result = (True, func(*args, **kwds))
      File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/multiprocessing/pool.py", line 44, in mapstar
        return list(map(*args))
      File "/Users/carol/Projects/jupyter-world/notebook/setupbase.py", line 473, in build_main
        if not self.should_run(name, target):
      File "/Users/carol/Projects/jupyter-world/notebook/setupbase.py", line 464, in should_run
        if mtime(source) > target_mtime:
      File "/Users/carol/Projects/jupyter-world/notebook/setupbase.py", line 308, in mtime
        return os.stat(path).st_mtime
    FileNotFoundError: [Errno 2] No such file or directory: '/Users/carol/Projects/jupyter-world/notebook/notebook/static/notebook/js/built/main.js'

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/Users/carol/Projects/jupyter-world/notebook/setup.py", line 199, in <module>
        main()
      File "/Users/carol/Projects/jupyter-world/notebook/setup.py", line 196, in main
        setup(**setup_args)
      File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/distutils/dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/Users/carol/Projects/jupyter-world/notebook/setupbase.py", line 549, in run
        log.warning("rebuilding js and css failed (not a problem)")
    AttributeError: module 'distutils.log' has no attribute 'warning'

    ----------------------------------------
Command "/Users/carol/.virtualenvs/mypy3/bin/python3.5 -c "import setuptools, tokenize;__file__='/Users/carol/Projects/jupyter-world/notebook/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" develop --no-deps" failed with error code 1 in /Users/carol/Projects/jupyter-world/notebook
```

**Comment by jdfreder** [comment]
^ I think that's a setup tools problem, we were able to get it working in person - but pip was still failing.  I don't think I wrote the offending line (it's not in the diff).

**Comment by blink1073** [comment]
Installed and ran locally, :+1: from me.

**Comment by jdfreder** [comment]
Thanks @blink1073 !

File: setupbase.py [hunk 3]
```diff
@@ -427,11 +434,11 @@ def finalize_options(self):
         self.force = bool(self.force)
 
     apps = ['notebook', 'tree', 'edit', 'terminal', 'auth']
-    targets = [ pjoin(static, app, 'js', 'main.min.js') for app in apps ]
+    targets = [ pjoin(static, app, 'js', 'built', 'main.min.js') for app in apps ]
     
     def sources(self, name):
         """Generator yielding .js sources that an application depends on"""
-        yield pjoin(static, name, 'js', 'main.js')
+        yield pjoin(static, name, 'js', 'built', 'main.js')
```
Comment (minrk) [inline setupbase.py]: This should now be main.min.js

**Comment by minrk** [comment]
The AttributeError there is fixed by #1125.

The FileNotFound is a real bug, and only shows up if you do `pip install -e` *again* after everything is built. It's a typo in `CompileJS.sources` where a `main.js` should be `main.min.js` (L441).

**Comment by jdfreder** [comment]
> The FileNotFound is a real bug, and only shows up if you do pip install -e again after everything is built. It's a typo in CompileJS.sources where a main.js should be main.min.js (L441).

Thanks, I actually forgot a couple to come back and update setupbase after making changes to the build later in the PR.  Last commit (252b0fc) should have aformentioned main name change, and a few other missing build files.

File: setupbase.py (hunk 3 above)
Comment (jdfreder) [inline setupbase.py]: :heavy_check_mark:

**Comment by minrk** [comment]
Great, thanks! I'm :+1: to merge this now, and get testing in the wild. Will check in at tonight's dev meeting.

**Comment by jdfreder** [comment]
From the meeting it sounds like the PR is in favor of merge - thanks for bringing it up @minrk !

**Comment by willingc** [comment]
Thanks @jdfreder :sunflower:

**Pull request merged**